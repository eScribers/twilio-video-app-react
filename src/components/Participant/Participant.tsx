import React from 'react';
import ParticipantInfo from '../ParticipantInfo/ParticipantInfo';
import ParticipantTracks from '../ParticipantTracks/ParticipantTracks';
import { Participant as IParticipant } from 'twilio-video';

interface ParticipantProps {
  participant: IParticipant;
  disableAudio?: boolean;
  enableScreenShare?: boolean;
  onClick: () => void;
  isSelected: boolean;
  gridView?: boolean;
}

export default function Participant({
  participant,
  disableAudio,
  enableScreenShare,
  onClick,
  isSelected,
  gridView,
}: ParticipantProps) {
  return (
    <ParticipantInfo participant={participant} onClick={onClick} isSelected={isSelected} gridView={gridView}>
      <ParticipantTracks participant={participant} disableAudio={disableAudio} />
    </ParticipantInfo>
  );
}
