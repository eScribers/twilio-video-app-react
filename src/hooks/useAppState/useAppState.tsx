import { useContext } from 'react';
import { StateContext } from '../../state/AppStateProvider';

export function useAppState(): any {
  const context = useContext(StateContext);
  if (!context) {
    throw new Error('useAppState must be used within the AppStateProvider');
  }
  return context;
}
