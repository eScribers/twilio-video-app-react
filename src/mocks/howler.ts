window.HTMLMediaElement.prototype.load = () => {
  /* do nothing */
};
window.HTMLMediaElement.prototype.play = async () => {
  /* do nothing */
};
window.HTMLMediaElement.prototype.pause = () => {
  /* do nothing */
};
window.HTMLMediaElement.prototype.addTextTrack = (_kind, _label: string | undefined, _language: string | undefined) =>
  ({} as TextTrack);

export {};
