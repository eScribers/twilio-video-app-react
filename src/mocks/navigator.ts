const mockGetUserMedia = jest.fn(async () => {
  return new Promise<void>(resolve => {
    resolve();
  });
});

const mockEnumerateDevices = jest.fn(
  () =>
    new Promise((res, _rej) => {
      res([]);
    })
);

Object.defineProperty(navigator, 'mediaDevices', {
  value: {
    enumerateDevices: mockEnumerateDevices,
    getUserMedia: mockGetUserMedia,
    addEventListener: jest.fn(),
  },
});

export {};
