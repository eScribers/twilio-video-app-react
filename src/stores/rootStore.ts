import rootStore from './makeStore';

export default rootStore;
