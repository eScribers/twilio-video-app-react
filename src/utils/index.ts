import isPlainObject from 'is-plain-object';

export const isMobile = (() => {
  if (typeof navigator === 'undefined' || typeof navigator.userAgent !== 'string') {
    return false;
  }
  return /Mobile/.test(navigator.userAgent);
})();

// This function ensures that the user has granted the browser permission to use audio and video
// devices. If permission has not been granted, it will cause the browser to ask for permission
// for audio and video at the same time (as opposed to separate requests).
export function ensureMediaPermissions() {
  return navigator.mediaDevices
    .enumerateDevices()
    .then(devices => devices.every(device => !(device.deviceId && device.label)))
    .then(shouldAskForMediaPermissions => {
      if (shouldAskForMediaPermissions) {
        return navigator.mediaDevices
          .getUserMedia({ audio: true, video: true })
          .then(mediaStream => mediaStream.getTracks().forEach(track => track.stop()));
      }
    });
}

// Recursively removes any object keys with a value of undefined
export function removeUndefineds<T>(obj: T): T {
  if (!isPlainObject(obj)) return obj;

  const target: { [name: string]: any } = {};

  for (const key in obj) {
    const val = obj[key];
    if (typeof val !== 'undefined') {
      target[key] = removeUndefineds(val);
    }
  }

  return target as T;
}

export function detectBrowser() {
  // Get the user-agent string
  let userAgentString = navigator.userAgent;

  // Detect Chrome
  let chromeAgent = userAgentString.indexOf('Chrome') > -1;

  // Detect Internet Explorer
  let IExplorerAgent = userAgentString.indexOf('MSIE') > -1 || userAgentString.indexOf('rv:') > -1;

  // Detect Firefox
  let firefoxAgent = userAgentString.indexOf('Firefox') > -1;

  // Detect Safari
  let safariAgent = userAgentString.indexOf('Safari') > -1;

  // Discard Safari since it also matches Chrome
  if (chromeAgent && safariAgent) safariAgent = false;

  // Detect Opera
  let operaAgent = userAgentString.indexOf('OP') > -1;

  // Discard Chrome since it also matches Opera
  if (chromeAgent && operaAgent) chromeAgent = false;

  if (safariAgent) return 'safariAgent';
  if (chromeAgent) return 'chromeAgent';
  if (IExplorerAgent) return 'IExplorerAgent';
  if (operaAgent) return 'operaAgent';
  if (firefoxAgent) return 'firefoxAgent';
  return 'no detected browser';
}

export const sleep = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms));
};
