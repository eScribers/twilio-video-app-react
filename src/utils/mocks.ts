import {
  VideoTrack,
  Track,
  NetworkQualityLevel,
  NetworkQualityStats,
  LocalParticipant,
  LocalTrack,
  TwilioError,
  LocalTrackPublication,
  EncodingParameters,
  NetworkQualityConfiguration,
  LocalTrackPublishOptions,
  LocalVideoTrack,
  Participant,
} from 'twilio-video';
import { ROOM_STATE } from './displayStrings';

class baseParticipant {
  audioTracks = new Map();
  dataTracks = new Map();
  identity = 'testing@Reporter@1';
  networkQualityLevel = null;
  networkQualityStats = null;
  sid = Math.floor(Math.random() * 100000);
  state = 'todo';
  tracks = new Map();
  videoTracks = new Map();
  addListener = (_event: string, _listener = () => {}) => this;
  once = (_event: string, ..._args: any[]) => this;
  prependListener = (_event: string, ..._args: any[]) => this;
  prependOnceListener = (_event: string, ..._args: any[]) => this;
  removeListener = (_event: string, ..._args: any[]) => this;
  off = (_event: string, ..._args: any[]) => this;
  removeAllListeners = (_event: string) => this;
  setMaxListeners = (_n: number) => this;
  getMaxListeners = () => {
    return 2;
  };
  listeners = (_event: string) => {
    return [() => {}];
  };
  rawListeners = (_event: string) => {
    return [() => {}];
  };
  emit = (_event: string, ..._args: any[]) => true;
  eventNames = () => [];
  listenerCount = (_type: string) => 1;
}

export class mockParticipant extends baseParticipant implements Participant {
  identity = 'remote@Reporter@2';
  on(event: ROOM_STATE.DISCONNECTED, listener: (participant: this) => void): this;
  on(
    event: 'networkQualityLevelChanged',
    listener: (networkQualityLevel: NetworkQualityLevel, networkQualityStats: NetworkQualityStats) => void
  ): this;
  on(event: ROOM_STATE.RECONNECTED, listener: (participant: this) => void): this;
  on(event: ROOM_STATE.RECONNECTING, listener: (participant: this) => void): this;
  on(event: 'trackDimensionsChanged', listener: (track: VideoTrack) => void): this;
  on(event: 'trackStarted', listener: (track: Track) => void): this;
  on(_event: string, _listener: (...args: any[]) => void) {
    return this;
  }

  constructor(name?: string, role?: string, id?: number, sid?: number) {
    super();
    if (name) this.identity = `${name}@${role}@${id}`;
    console.log(this.identity);

    if (sid) {
      this.sid = sid;
    }
  }
}

export class mockLocalParticipant extends baseParticipant implements LocalParticipant {
  identity = 'local@Reporter@1';
  audioTracks = new Map();
  dataTracks = new Map();
  tracks = new Map();
  videoTracks = new Map();
  signalingRegion = 'todo';

  constructor(name?: string, role?: string, id?: number, userId?: number) {
    super();
    if (name) this.identity = [name, role, id, userId].filter(x => x).join('@');
  }

  publishTrack = async (_track: LocalTrack | MediaStreamTrack, _options?: LocalTrackPublishOptions) =>
    ({} as LocalTrackPublication);
  publishTracks = async (_tracks: Array<LocalTrack | MediaStreamTrack>): Promise<LocalTrackPublication[]> => [];
  setNetworkQualityConfiguration = (_networkQualityConfiguration: NetworkQualityConfiguration) => this;
  setParameters = (_encodingParameters?: EncodingParameters | null) => this;
  unpublishTrack = (_track: LocalTrack | MediaStreamTrack): LocalTrackPublication | null => null;
  unpublishTracks = (_tracks: Array<LocalTrack | MediaStreamTrack>): LocalTrackPublication[] => [];
  on(event: ROOM_STATE.DISCONNECTED, listener: (participant: this) => void): this;
  on(event: 'trackDimensionsChanged', listener: (track: LocalVideoTrack) => void): this;
  on(event: 'trackDisabled', listener: (track: LocalTrack) => void): this;
  on(event: 'trackEnabled', listener: (track: LocalTrack) => void): this;
  on(event: 'trackPublicationFailed', listener: (error: TwilioError, track: LocalTrack) => void): this;
  on(event: 'trackPublished', listener: (publication: LocalTrackPublication) => void): this;
  on(event: 'trackStarted', listener: (track: LocalTrack) => void): this;
  on(event: 'trackStopped', listener: (track: LocalTrack) => void): this;
  on(_event: string, _listener: (...args: any[]) => void) {
    return this;
  }
}
