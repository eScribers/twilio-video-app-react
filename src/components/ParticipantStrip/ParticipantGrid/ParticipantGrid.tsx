import React from 'react';
import { observer } from 'mobx-react-lite';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import { useEffect, useState } from 'react';
import Grid from '@material-ui/core/Grid';
import Participant from '../../Participant/Participant';
import rootStore from '../../../stores';

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
    display: 'flex',
    [theme.breakpoints.down('xs')]: {
      flexWrap: 'wrap',
    },
    flexWrap: 'nowrap',
  },
  paper: {
    padding: theme.spacing(1),
    textAlign: 'center',
    color: theme.palette.text.secondary,
  },
  scrollable: {
    overflow: 'auto',
  },
}));

export interface ParticipantGridProps {
  viewMode: string;
}

const ParticipantGrid = observer(({ viewMode }: ParticipantGridProps) => {
  const { participantsStore } = rootStore;
  const [currViewMode, setCurrViewMode] = useState('');
  const [lgState, setLgState] = useState<any>(3);
  const [mdState, setMdState] = useState<any>(4);
  const { sortedParticipants, selectedParticipant, participant: localParticipant } = participantsStore;
  const classes = useStyles();

  const dominantIdentity = participantsStore.dominantSpeaker;

  useEffect(() => {
    if (currViewMode !== viewMode) {
      if (viewMode.includes('2 column')) {
        setMdState(6);
        setLgState(6);
      } else if (viewMode.includes('3 column')) {
        setMdState(4);
        setLgState(4);
      } else if (viewMode.includes('4 column')) {
        setMdState(3);
        setLgState(3);
      } else {
        setMdState(4);
        setLgState(3);
      }
      setCurrViewMode(viewMode);
    }
  }, [viewMode, currViewMode]);

  if (!localParticipant) return null;
  return (
    <div className={classes.root}>
      <Grid container spacing={3} className={classes.scrollable}>
        <Grid item xs={12} sm={6} md={mdState} lg={lgState}>
          <Paper className={classes.paper}>
            <Participant
              participant={localParticipant}
              isSelected={selectedParticipant === localParticipant.identity}
              onClick={() => participantsStore.setSelectedParticipant(localParticipant.identity)}
            />
          </Paper>
        </Grid>
        {sortedParticipants.map((participant: any) => (
          <Grid key={participant.sid} item xs={12} sm={6} md={mdState} lg={lgState}>
            <Paper className={classes.paper}>
              <Participant
                key={participant.sid}
                participant={participant}
                isSelected={selectedParticipant === participant}
                isDominantSpeaker={dominantIdentity === participant.identity}
                userIsSilenced={!!participantsStore.isSilenced}
                onClick={() => participantsStore.setSelectedParticipant(participant.identity)}
              />
            </Paper>
          </Grid>
        ))}
      </Grid>
    </div>
  );
});

export default ParticipantGrid;
