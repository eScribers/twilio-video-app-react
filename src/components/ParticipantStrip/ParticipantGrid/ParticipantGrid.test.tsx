import rootStore, { RootStore } from '../../../stores/makeStore';
import React from 'react';
import ParticipantGrid from './ParticipantGrid';
import { shallow } from 'enzyme';
import { mockLocalParticipant, mockParticipant } from '../../../utils/mocks';
import { useAppState } from '../../../hooks/useAppState/useAppState';
import { act } from '@testing-library/react';
import Participant from '../../Participant/Participant';

jest.mock('../../../hooks/useAppState/useAppState');

const mockUseAppState = useAppState as jest.Mock<any>;

jest.mock('../../../stores', () => {
  return {
    __esModule: true, // this property makes it work
    default: rootStore,
  };
});

const mockLocalTrack = {
  kind: 'audioinput',
  mediaStreamTrack: {
    label: 'mock local audio track',
    getSettings: () => ({ deviceId: '234' }),
  },
  restart: jest.fn(),
  on: jest.fn(),
  off: jest.fn(),
  enable: jest.fn(),
};
mockUseAppState.mockImplementation(() => ({ activeSinkId: '' }));
// const mockLocalParticipant = {};

describe('the ParticipantStrip component', () => {
  beforeEach(() => {
    let newStore = new RootStore();
    rootStore.participantsStore = newStore.participantsStore;
    rootStore.roomsStore = newStore.roomsStore;
  });

  it('should correctly render ParticipantInfo components', () => {
    const localParticipant = new mockLocalParticipant('local', 'Reporter', 1);
    rootStore.participantsStore.localParticipant?.setParticipant({ ...localParticipant, sid: '123' });
    rootStore.participantsStore.setAudioTrack(mockLocalTrack);
    const participant1 = new mockParticipant('participant1', 'Reporter', 1, 1);
    const participant2 = new mockParticipant('participant2', 'Reporter', 2, 2);
    rootStore.participantsStore.setParticipants([participant1, participant2]);

    const wrapper = shallow(<ParticipantGrid viewMode={'grid 3x3'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should add the isSelected prop to the local participant when it is selected', () => {
    const localParticipant = new mockLocalParticipant('local', 'Reporter', 1);
    rootStore.participantsStore.localParticipant?.setParticipant({ ...localParticipant, sid: '123' });
    rootStore.participantsStore.setAudioTrack(mockLocalTrack);
    const participant1 = new mockParticipant('participant1', 'Reporter', 1, 1);
    const participant2 = new mockParticipant('participant2', 'Reporter', 2, 2);
    rootStore.participantsStore.setParticipants([participant1, participant2]);

    const wrapper = shallow(<ParticipantGrid viewMode={'grid 3x3'} />);
    act(() => {
      wrapper
        .find(Participant)
        .at(0)
        .simulate('click');
    });
    expect(rootStore.participantsStore.selectedParticipant).toBe(localParticipant.identity);
  });

  it('should add the isSelected prop to the first remote participant when it is selected', () => {
    const localParticipant = new mockLocalParticipant('local', 'Reporter', 1);
    rootStore.participantsStore.localParticipant?.setParticipant({ ...localParticipant, sid: '123' });
    rootStore.participantsStore.setAudioTrack(mockLocalTrack);
    const participant1 = new mockParticipant('participant1', 'Reporter', 1, 1);
    const participant2 = new mockParticipant('participant2', 'Reporter', 2, 2);
    rootStore.participantsStore.setParticipants([participant1, participant2]);

    const wrapper = shallow(<ParticipantGrid viewMode={'grid 3x3'} />);
    act(() => {
      wrapper
        .find(Participant)
        .at(1)
        .simulate('click');
    });
    expect(rootStore.participantsStore.selectedParticipant).toBe(participant1.identity);
  });
});
