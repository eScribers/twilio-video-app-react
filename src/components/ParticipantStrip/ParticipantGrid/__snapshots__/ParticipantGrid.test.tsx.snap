// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ParticipantStrip component should correctly render ParticipantInfo components 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-scrollable-3"
    container={true}
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={3}
      md={4}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <Memo(wrappedComponent)
          isSelected={false}
          onClick={[Function]}
          participant={
            Object {
              "addListener": [Function],
              "audioTracks": Map {},
              "dataTracks": Map {},
              "emit": [Function],
              "eventNames": [Function],
              "getMaxListeners": [Function],
              "identity": "local@Reporter@1",
              "listenerCount": [Function],
              "listeners": [Function],
              "networkQualityLevel": null,
              "networkQualityStats": null,
              "off": [Function],
              "once": [Function],
              "prependListener": [Function],
              "prependOnceListener": [Function],
              "publishTrack": [Function],
              "publishTracks": [Function],
              "rawListeners": [Function],
              "removeAllListeners": [Function],
              "removeListener": [Function],
              "setMaxListeners": [Function],
              "setNetworkQualityConfiguration": [Function],
              "setParameters": [Function],
              "sid": "123",
              "signalingRegion": "todo",
              "state": "todo",
              "tracks": Map {},
              "unpublishTrack": [Function],
              "unpublishTracks": [Function],
              "videoTracks": Map {},
            }
          }
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      lg={3}
      md={4}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <Memo(wrappedComponent)
          isDominantSpeaker={false}
          isSelected={false}
          key="1"
          onClick={[Function]}
          participant={
            mockParticipant {
              "addListener": [Function],
              "audioTracks": Map {},
              "dataTracks": Map {},
              "emit": [Function],
              "eventNames": [Function],
              "getMaxListeners": [Function],
              "identity": "participant1@Reporter@1",
              "listenerCount": [Function],
              "listeners": [Function],
              "networkQualityLevel": null,
              "networkQualityStats": null,
              "off": [Function],
              "once": [Function],
              "prependListener": [Function],
              "prependOnceListener": [Function],
              "rawListeners": [Function],
              "removeAllListeners": [Function],
              "removeListener": [Function],
              "setMaxListeners": [Function],
              "sid": 1,
              "state": "todo",
              "tracks": Map {},
              "videoTracks": Map {},
            }
          }
          userIsSilenced={false}
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="2"
      lg={3}
      md={4}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <Memo(wrappedComponent)
          isDominantSpeaker={false}
          isSelected={false}
          key="2"
          onClick={[Function]}
          participant={
            mockParticipant {
              "addListener": [Function],
              "audioTracks": Map {},
              "dataTracks": Map {},
              "emit": [Function],
              "eventNames": [Function],
              "getMaxListeners": [Function],
              "identity": "participant2@Reporter@2",
              "listenerCount": [Function],
              "listeners": [Function],
              "networkQualityLevel": null,
              "networkQualityStats": null,
              "off": [Function],
              "once": [Function],
              "prependListener": [Function],
              "prependOnceListener": [Function],
              "rawListeners": [Function],
              "removeAllListeners": [Function],
              "removeListener": [Function],
              "setMaxListeners": [Function],
              "sid": 2,
              "state": "todo",
              "tracks": Map {},
              "videoTracks": Map {},
            }
          }
          userIsSilenced={false}
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
